File: scripts/attribute/杀敌时加属性.lua
local ____lualib = require("lualib_bundle")
local __TS__Class = ____lualib.__TS__Class
local __TS__SourceMapTraceBack = ____lualib.__TS__SourceMapTraceBack
__TS__SourceMapTraceBack(debug.getinfo(1).short_src, {["6"] = 1,["7"] = 1,["8"] = 2,["9"] = 2,["10"] = 3,["11"] = 3,["12"] = 4,["13"] = 4,["14"] = 49,["15"] = 49,["16"] = 49,["18"] = 51,["19"] = 52,["22"] = 55,["23"] = 56,["24"] = 58,["25"] = 59,["26"] = 59,["27"] = 59,["28"] = 59,["29"] = 60,["30"] = 61,["32"] = 64,["33"] = 65,["34"] = 67,["35"] = 68,["37"] = 71,["38"] = 72,["40"] = 76,["41"] = 77,["42"] = 78,["43"] = 79,["45"] = 81,["46"] = 82,["47"] = 82,["48"] = 82,["49"] = 82,["50"] = 83,["52"] = 88,["53"] = 89,["54"] = 90,["55"] = 91,["57"] = 93,["58"] = 94,["59"] = 94,["60"] = 94,["61"] = 94,["62"] = 95,["64"] = 100,["65"] = 101,["66"] = 102,["67"] = 103,["69"] = 105,["70"] = 106,["71"] = 106,["72"] = 106,["73"] = 106,["74"] = 106,["76"] = 111,["77"] = 112,["78"] = 113,["79"] = 114,["81"] = 116,["82"] = 117,["83"] = 117,["84"] = 117,["85"] = 117,["86"] = 117,["88"] = 121,["89"] = 122,["90"] = 123,["91"] = 124,["93"] = 126,["94"] = 127,["95"] = 127,["96"] = 127,["97"] = 127,["98"] = 127,["100"] = 130,["101"] = 131,["102"] = 131,["103"] = 131,["104"] = 131,["105"] = 132,["107"] = 135,["108"] = 136,["109"] = 137,["110"] = 138,["112"] = 140,["113"] = 141,["114"] = 141,["115"] = 141,["116"] = 141,["119"] = 51,["120"] = 50});
local ____exports = {}
local ____AttributeUtil = require("solar.solar-common.util.system.AttributeUtil")
local AttributeUtil = ____AttributeUtil.default
local ____UnitStateUtil = require("solar.solar-common.util.unit.UnitStateUtil")
local UnitStateUtil = ____UnitStateUtil.default
local ____PlayerUtil = require("solar.solar-common.util.game.PlayerUtil")
local PlayerUtil = ____PlayerUtil.default
local ____DataBase = require("solar.solar-common.common.DataBase")
local DataBase = ____DataBase.default
____exports.default = __TS__Class()
local _____6740_654C_52A0_5C5E_6027 = ____exports.default
_____6740_654C_52A0_5C5E_6027.name = "杀敌加属性"
function _____6740_654C_52A0_5C5E_6027.prototype.____constructor(self)
    se:onUnitDeath(function(e)
        if not e.hasKillingUnit then
            return
        end
        local triggerunit = e.trigUnit
        local killunit = e.killingUnit
        local attribute = AttributeUtil:getUnitAttribute(killunit)
        if IsUnitEnemy(
            triggerunit,
            GetOwningPlayer(killunit)
        ) and IsHandle(killunit) and attribute then
            if attribute.reply_life_on_kill then
                UnitStateUtil:addLife(killunit, attribute.reply_life_on_kill)
            end
            local add_max_life = attribute.add_max_life_on_kill or 0
            local add_max_life_on_kill_p = attribute.add_max_life_on_kill_p or 0
            if add_max_life_on_kill_p ~= 0 then
                add_max_life = add_max_life * (add_max_life_on_kill_p + 1)
            end
            if add_max_life > 0 then
                UnitStateUtil:addMaxLifeAndLife(killunit, add_max_life)
            end
            local add_attack_on_kill = attribute.add_attack_on_kill or 0
            local add_attack_on_kill_p = attribute.add_attack_on_kill_p or 0
            if add_attack_on_kill_p ~= 0 then
                add_attack_on_kill = add_attack_on_kill * (add_attack_on_kill_p + 1)
            end
            if add_attack_on_kill > 0 then
                local attack = GetUnitState(
                    killunit,
                    ConvertUnitState(18)
                )
                UnitStateUtil:setDamageBase(killunit, add_attack_on_kill + attack)
            end
            local add_def_on_kill = attribute.add_def_on_kill or 0
            local add_def_on_kill_p = attribute.add_def_on_kill_p or 0
            if add_def_on_kill_p ~= 0 then
                add_def_on_kill = add_def_on_kill * (add_def_on_kill_p + 1)
            end
            if add_def_on_kill > 0 then
                local def = GetUnitState(
                    killunit,
                    ConvertUnitState(32)
                )
                UnitStateUtil:setArmor(killunit, add_def_on_kill + def)
            end
            local add_strength_on_kill = attribute.add_strength_on_kill or 0
            local add_strength_on_kill_p = attribute.add_strength_on_kill_p or 0
            if add_strength_on_kill_p ~= 0 then
                add_strength_on_kill = add_strength_on_kill * (add_strength_on_kill_p + 1)
            end
            if add_strength_on_kill > 0 then
                SetHeroStr(
                    killunit,
                    GetHeroStr(killunit, false) + add_strength_on_kill,
                    true
                )
            end
            local add_agility_on_kill = attribute.add_agility_on_kill or 0
            local add_agility_on_kill_p = attribute.add_agility_on_kill_p or 0
            if add_agility_on_kill_p ~= 0 then
                add_agility_on_kill = add_agility_on_kill * (add_agility_on_kill_p + 1)
            end
            if add_agility_on_kill > 0 then
                SetHeroAgi(
                    killunit,
                    GetHeroAgi(killunit, false) + add_agility_on_kill,
                    true
                )
            end
            local add_intelligence_on_kill = attribute.add_intelligence_on_kill or 0
            local add_intelligence_on_kill_p = attribute.add_intelligence_on_kill_p or 0
            if add_intelligence_on_kill_p ~= 0 then
                add_intelligence_on_kill = add_intelligence_on_kill * (add_intelligence_on_kill_p + 1)
            end
            if add_intelligence_on_kill > 0 then
                SetHeroInt(
                    killunit,
                    GetHeroInt(killunit, false) + add_intelligence_on_kill,
                    true
                )
            end
            if attribute.add_killCount_on_kill and attribute.add_killCount_on_kill > 0 then
                local solarData = DataBase:getPlayerSolarData(
                    GetOwningPlayer(killunit),
                    true
                )
                solarData.killCount = (solarData.killCount or 0) + attribute.add_killCount_on_kill
            end
            local gold = attribute.add_gold_on_kill or 0
            local basics_gold = attribute.earn_gold_p or 0
            if gold ~= 0 and basics_gold ~= 0 then
                gold = gold * (basics_gold + 1)
            end
            if gold > 0 then
                PlayerUtil:addGoldState(
                    GetOwningPlayer(killunit),
                    gold
                )
            end
        end
    end)
end
return ____exports

File: src/attribute/杀敌时加属性.ts
import AttributeUtil from '@/AttributeUtil';
import UnitStateUtil from '@/UnitStateUtil';
import PlayerUtil from "@/PlayerUtil";
import DataBase from "@/DataBase";

/**
 * 自定义的杀敌加生命
 * 杀敌加生命增幅
 */
declare global {
    interface AppAttribute {
        /**杀敌加生命*/
        /**杀敌时恢复当前生命值*/
        reply_life_on_kill?: number
        /**单位每击杀一个敌人增加生命值上限的量。不受杀敌倍率的影响，初始值为0*/
        add_max_life_on_kill?: number
        /**杀敌加生命增幅*/
        /**按比例增幅单位的杀敌加生命值上限数值，初始值为0%。当不为0时，以(+XX%)的方式现在在属性面板的对应属性后面*/
        add_max_life_on_kill_p?: number
        /**攻击*/
        add_attack_on_kill?: number
        add_attack_on_kill_p?: number

        /**护甲*/
        add_def_on_kill?: number
        add_def_on_kill_p?: number

        /**力量*/
        add_strength_on_kill?: number
        add_strength_on_kill_p?: number

        /**敏捷*/
        add_agility_on_kill?: number
        add_agility_on_kill_p?: number

        /**智力*/
        add_intelligence_on_kill?: number
        add_intelligence_on_kill_p?: number


        /**杀敌金币*/
        add_gold_on_kill?: number
        /**杀敌加额外的杀敌数*/
        add_killCount_on_kill?: number


    }
}
export default class 杀敌加属性 {
    constructor() {
        se.onUnitDeath((e) => {
            if (!e.hasKillingUnit) {
                return
            }
            let triggerunit = e.trigUnit;
            let killunit = e.killingUnit;

            let attribute: AppAttribute = AttributeUtil.getUnitAttribute(killunit);
            if (IsUnitEnemy(triggerunit, GetOwningPlayer(killunit)) && IsHandle(killunit) && attribute) {
                if (attribute.reply_life_on_kill) {
                    UnitStateUtil.addLife(killunit, attribute.reply_life_on_kill)
                }
                //获得初始的
                let add_max_life = attribute.add_max_life_on_kill || 0
                let add_max_life_on_kill_p = attribute.add_max_life_on_kill_p || 0
                //如果单位的杀敌加生命增幅存在,则修改储存的基础杀敌加生命以及刷新生命增幅
                if (add_max_life_on_kill_p != 0) {
                    add_max_life = add_max_life * (add_max_life_on_kill_p + 1)
                }
                // print("增加生命上限"+add_max_life)
                if (add_max_life > 0) {
                    UnitStateUtil.addMaxLifeAndLife(killunit, add_max_life)
                }

                //杀敌加攻击
                let add_attack_on_kill = attribute.add_attack_on_kill || 0
                let add_attack_on_kill_p = attribute.add_attack_on_kill_p || 0
                if (add_attack_on_kill_p != 0) {
                    add_attack_on_kill = add_attack_on_kill * (add_attack_on_kill_p + 1)
                }
                if (add_attack_on_kill > 0) {
                    let attack = GetUnitState(killunit, ConvertUnitState(0x12))
                    UnitStateUtil.setDamageBase(killunit, (add_attack_on_kill + attack))
                    // AttributeUtil.getUnitAttribute(killunit).add_attack_on_kill
                }

                //杀敌加护甲
                let add_def_on_kill = attribute.add_def_on_kill || 0
                let add_def_on_kill_p = attribute.add_def_on_kill_p || 0
                if (add_def_on_kill_p != 0) {
                    add_def_on_kill = add_def_on_kill * (add_def_on_kill_p + 1)
                }
                if (add_def_on_kill > 0) {
                    let def = GetUnitState(killunit, ConvertUnitState(0x20))
                    UnitStateUtil.setArmor(killunit, (add_def_on_kill + def))
                    // AttributeUtil.getUnitAttribute(killunit).add_attack_on_kill
                }

                //杀敌加力量
                let add_strength_on_kill = attribute.add_strength_on_kill || 0
                let add_strength_on_kill_p = attribute.add_strength_on_kill_p || 0
                if (add_strength_on_kill_p != 0) {
                    add_strength_on_kill = add_strength_on_kill * (add_strength_on_kill_p + 1)
                }
                if (add_strength_on_kill > 0) {
                    SetHeroStr(killunit, (GetHeroStr(killunit, false) + add_strength_on_kill), true)
                }


                //杀敌加敏捷
                let add_agility_on_kill = attribute.add_agility_on_kill || 0
                let add_agility_on_kill_p = attribute.add_agility_on_kill_p || 0
                if (add_agility_on_kill_p != 0) {
                    add_agility_on_kill = add_agility_on_kill * (add_agility_on_kill_p + 1)
                }
                if (add_agility_on_kill > 0) {
                    SetHeroAgi(killunit, (GetHeroAgi(killunit, false) + add_agility_on_kill), true)
                }

                //杀敌加智力
                let add_intelligence_on_kill = attribute.add_intelligence_on_kill || 0
                let add_intelligence_on_kill_p = attribute.add_intelligence_on_kill_p || 0
                if (add_intelligence_on_kill_p != 0) {
                    add_intelligence_on_kill = add_intelligence_on_kill * (add_intelligence_on_kill_p + 1)
                }
                if (add_intelligence_on_kill > 0) {
                    SetHeroInt(killunit, (GetHeroInt(killunit, false) + add_intelligence_on_kill), true)
                }

                if (attribute.add_killCount_on_kill && attribute.add_killCount_on_kill > 0) {
                    let solarData = DataBase.getPlayerSolarData(GetOwningPlayer(killunit), true);
                    solarData.killCount = (solarData.killCount || 0) + attribute.add_killCount_on_kill
                }

                let gold = attribute.add_gold_on_kill || 0
                let basics_gold = attribute.earn_gold_p || 0
                if (gold != 0 && basics_gold != 0) {
                    gold = gold * (basics_gold + 1)
                }
                if (gold > 0) {
                    PlayerUtil.addGoldState(GetOwningPlayer(killunit), gold)
                }


            }
        })

    }
}